File: mvpcore/src/main/java/biz/dealnote/mvp/compat/ViewHostDelegate.java
package biz.dealnote.mvp.compat;

import android.content.Context;
import android.os.Bundle;
import android.support.annotation.NonNull;
import android.support.annotation.Nullable;
import android.support.v4.app.LoaderManager;
import android.support.v4.content.Loader;

import java.lang.ref.WeakReference;
import java.util.ArrayList;
import java.util.List;

import biz.dealnote.mvp.core.IMvpView;
import biz.dealnote.mvp.core.IPresenter;
import biz.dealnote.mvp.core.IPresenterFactory;
import biz.dealnote.mvp.core.PresenterAction;

/**
 * Created by ruslan.kolbasa on 08.09.2016.
 * mvpcore
 */
public class ViewHostDelegate<P extends IPresenter<V>, V extends IMvpView> {

    private static final String SAVE_PRESENTER_STATE = "save-presenter-state";
    private static final int LOADER_ID = 101;

    private Bundle mLastKnownPresenterState;

    private boolean mViewCreated;

    private P mPresenter;
    private WeakReference<V> mViewReference;

    public void onCreate(@NonNull final Context context,
                         @NonNull V viewHost,
                         @NonNull final IFactoryProvider<P, V> factoryProvider,
                         @NonNull LoaderManager loaderManager,
                         @Nullable Bundle savedInstanceState) {
        this.mViewReference = new WeakReference<>(viewHost);

        if (savedInstanceState != null) {
            this.mLastKnownPresenterState = savedInstanceState.getBundle(SAVE_PRESENTER_STATE);
        }

        final Context app = context.getApplicationContext();
        SimplePresenterLoader<P, V> loader = (SimplePresenterLoader<P, V>) loaderManager.initLoader(LOADER_ID, mLastKnownPresenterState, new LoaderManager.LoaderCallbacks<P>() {
            @NonNull
            @Override
            public Loader<P> onCreateLoader(int id, Bundle args) {
                return new SimplePresenterLoader<>(app, factoryProvider.getPresenterFactory(args));
            }

            @Override
            public void onLoadFinished(@NonNull Loader<P> loader, P data) {

            }

            @Override
            public void onLoaderReset(@NonNull Loader<P> loader) {
                mPresenter = null;
            }
        });

        mPresenter = loader.get();
        mPresenter.attachViewHost(mViewReference.get());

        List<PresenterAction<P, V>> copy = new ArrayList<>(onReceivePresenterActions);
        onReceivePresenterActions.clear();
        for(PresenterAction<P, V> action : copy){
            action.call(mPresenter);
        }
    }

    public void onDestroy() {
        mPresenter.detachViewHost();
        mViewReference = new WeakReference<>(null);
    }

    public void onViewCreated() {
        if (mViewCreated) {
            return;
        }

        mViewCreated = true;
        mPresenter.createView(mViewReference.get());
    }

    public void onDestroyView() {
        mViewCreated = false;
        mPresenter.destroyView();
    }

    private List<PresenterAction<P, V>> onReceivePresenterActions = new ArrayList<>();

    public void callPresenter(PresenterAction<P, V> action){
        if(isPresenterPrepared()){
            action.call(getPresenter());
        }
    }

    public void postPrenseterReceive(PresenterAction<P, V> action){
        if(isPresenterPrepared()){
            action.call(getPresenter());
        } else {
            onReceivePresenterActions.add(action);
        }
    }

    public P getPresenter() {
        return mPresenter;
    }

    public void onResume() {
        mPresenter.resumeView();
    }

    public void onPause() {
        mPresenter.pauseView();
    }

    public boolean isPresenterPrepared() {
        return mPresenter != null;
    }

    public void onSaveInstanceState(Bundle outState) {
        if (mPresenter != null) {
            mLastKnownPresenterState = new Bundle();
            mPresenter.saveState(outState);
        }

        outState.putBundle(SAVE_PRESENTER_STATE, mLastKnownPresenterState);
    }

    public interface IFactoryProvider<P extends IPresenter<V>, V extends IMvpView> {
        IPresenterFactory<P> getPresenterFactory(@Nullable Bundle saveInstanceState);
    }
}

File: mvpcore/src/main/java/biz/dealnote/mvp/compat/ViewHostDelegate.kt
package biz.dealnote.mvp.compat

import android.content.Context
import android.os.Bundle
import android.support.v4.app.LoaderManager
import android.support.v4.content.Loader
import biz.dealnote.mvp.core.IMvpView
import biz.dealnote.mvp.core.IPresenter
import biz.dealnote.mvp.core.IPresenterFactory
import biz.dealnote.mvp.core.PresenterAction
import java.lang.ref.WeakReference
import java.util.*

/**
 * Created by ruslan.kolbasa on 08.09.2016.
 * mvpcore
 */
class ViewHostDelegate<P : IPresenter<V>, V : IMvpView> {

    private var lastKnownPresenterState: Bundle? = null

    private var viewCreated: Boolean = false

    var presenter: P? = null

    private var viewReference: WeakReference<V?> = WeakReference(null)

    private val onReceivePresenterActions = ArrayList<PresenterAction<P, V>>()

    val isPresenterPrepared: Boolean
        get() = presenter != null

    fun onCreate(context: Context,
                 view: V,
                 factoryProvider: IFactoryProvider<P, V>,
                 loaderManager: LoaderManager,
                 savedInstanceState: Bundle?) {
        this.viewReference = WeakReference(view)

        if (savedInstanceState != null) {
            this.lastKnownPresenterState = savedInstanceState.getBundle(SAVE_PRESENTER_STATE)
        }

        val app = context.applicationContext
        val loader = loaderManager.initLoader(LOADER_ID, lastKnownPresenterState, object : LoaderManager.LoaderCallbacks<P> {
            override fun onCreateLoader(id: Int, args: Bundle?): Loader<P> {
                return SimplePresenterLoader(app, factoryProvider.getPresenterFactory(args))
            }

            override fun onLoadFinished(loader: Loader<P>, data: P) {

            }

            override fun onLoaderReset(loader: Loader<P>) {
                presenter = null
            }
        })

        presenter = (loader as SimplePresenterLoader<P, V>).get()
        presenter?.run {
            attachViewHost(view)

            val copy = ArrayList(onReceivePresenterActions)
            onReceivePresenterActions.clear()
            for (action in copy) {
                action.call(this)
            }
        }
    }

    fun onDestroy() {
        viewReference = WeakReference(null)
        presenter?.detachViewHost()
    }

    fun onViewCreated() {
        if (viewCreated) {
            return
        }

        viewCreated = true
        presenter?.createView(viewReference.get()!!)
    }

    fun onDestroyView() {
        viewCreated = false
        presenter?.destroyView()
    }

    fun callPresenter(action: PresenterAction<P, V>) {
        presenter?.run {
            action.call(this)
        }
    }

    fun postPrenseterReceive(action: PresenterAction<P, V>) {
        presenter?.run {
            action.call(this)
        } ?: run {
            onReceivePresenterActions.add(action)
        }
    }

    fun onResume() {
        presenter?.resumeView()
    }

    fun onPause() {
        presenter?.pauseView()
    }

    fun onSaveInstanceState(outState: Bundle) {
        presenter?.run {
            lastKnownPresenterState = Bundle()
            saveState(outState)
        }

        outState.putBundle(SAVE_PRESENTER_STATE, lastKnownPresenterState)
    }

    interface IFactoryProvider<P : IPresenter<V>, V : IMvpView> {
        fun getPresenterFactory(saveInstanceState: Bundle?): IPresenterFactory<P>
    }

    companion object {
        private const val SAVE_PRESENTER_STATE = "save-presenter-state"
        private const val LOADER_ID = 101
    }
}